common/download.cpp

```diff
@@ -486,17 +488,36 @@ static void print_progress(size_t current, size_t total) {
         return;
     }
 
+    static std::mutex mutex;
+    static std::map<std::thread::id, int> lines;
```

Comment:
This looks like a local variable at first glance, but because it's `static`, that effectively make it a global variable.

IMO we should avoid global variable if possible. Probably a better way is to make a dedicated class to manage threads and implement the mutex as a member of the class.